pygradle-plugin/src/main/groovy/com/linkedin/gradle/python/plugin/WheelFirstPlugin.java

```diff
@@ -0,0 +1,86 @@
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.tasks.TaskContainer;
+
+import java.io.File;
+import java.util.Objects;
+
+import static com.linkedin.gradle.python.util.StandardTextValues.CONFIGURATION_BUILD_REQS;
+import static com.linkedin.gradle.python.util.StandardTextValues.CONFIGURATION_PYDOCS;
+import static com.linkedin.gradle.python.util.StandardTextValues.CONFIGURATION_PYTHON;
+import static com.linkedin.gradle.python.util.StandardTextValues.CONFIGURATION_SETUP_REQS;
+import static com.linkedin.gradle.python.util.StandardTextValues.CONFIGURATION_TEST;
+import static com.linkedin.gradle.python.util.StandardTextValues.CONFIGURATION_VENV;
+import static com.linkedin.gradle.python.util.StandardTextValues.CONFIGURATION_WHEEL;
+import static com.linkedin.gradle.python.util.StandardTextValues.TASK_INSTALL_SETUP_REQS;
+
+public class WheelFirstPlugin implements Plugin<Project> {
+    @Override
+    public void apply(Project project) {
+        TaskContainer tasks = project.getTasks();
+
+        File cacheDir = new File(project.getGradle().getGradleUserHomeDir(), "pygradle-wheel-cache");
+
+        project.getPlugins().withType(PythonPlugin.class, plugin -> {
+
+            SupportedWheelFormats supportedWheelFormats = new SupportedWheelFormats();
+            CachedBackedWheelCache wheelCache = new CachedBackedWheelCache(cacheDir, supportedWheelFormats);
+
+            FindAbiForCurrentPythonTask abiScript = tasks.create("findPythonAbi", FindAbiForCurrentPythonTask.class, it -> {
+                it.dependsOn(tasks.getByName(TASK_INSTALL_SETUP_REQS.getValue()));
+                it.setSupportedWheelFormat(supportedWheelFormats);
+                it.addPythonDetails(() -> ExtensionUtils.getPythonExtension(project).getDetails());
+            });
+
+            ParallelWheelGenerationTask parallelWheelTask = tasks.create("parallelWheels", ParallelWheelGenerationTask.class, it -> {
+                it.dependsOn(abiScript);
+                ConfigurationContainer configurations = project.getConfigurations();
+                FileCollection dependencies = configurations.getByName(CONFIGURATION_SETUP_REQS.getValue())
+                    .plus(configurations.getByName(CONFIGURATION_BUILD_REQS.getValue()))
+                    .plus(configurations.getByName(CONFIGURATION_PYTHON.getValue()))
+                    .plus(configurations.getByName(CONFIGURATION_WHEEL.getValue()))
+                    .plus(configurations.getByName(CONFIGURATION_PYDOCS.getValue()))
+                    .plus(configurations.getByName(CONFIGURATION_VENV.getValue()))
+                    .plus(configurations.getByName(CONFIGURATION_TEST.getValue()));
+
+                it.setFilesToConvert(dependencies);
+                it.setWheelCache(wheelCache);
+                it.setCacheDir(cacheDir);
+                it.dependsOn(tasks.getByName(TASK_INSTALL_SETUP_REQS.getValue()));
```

Comment:
Could these two also be `TASK_VENV_CREATE` instead of waiting for setup requirements?